source/Timer.cpp

```diff
@@ -0,0 +1,327 @@
+			actions[TimerTrigger::TIMEUP].Load(child);
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "reset")
+			actions[TimerTrigger::RESET].Load(child);
+		else
+			child.PrintTrace("Skipping unrecognized attribute:");
+
+	}
+}
+
+
+
+// Note: the Save() function can assume this is an instantiated Timer, not a template,
+// so the time to wait will be saved fully calculated, and with any elapsed time subtracted.
+void Timer::Save(DataWriter &out) const
+{
+	// If this Timer should no longer appear in-game, don't serialize it.
+	if(isComplete)
+		return;
+
+	int64_t timeRemaining = timeToWait - timeElapsed;
+	out.Write("timer");
+	out.BeginChild();
+	{
+		out.Write("time", timeRemaining);
+		if(system)
+			out.Write("system", system->Name());
+		else if(!systems.IsEmpty())
+		{
+			out.Write("system");
+			systems.Save(out);
+		}
+		if(requireIdle)
+		{
+			out.Write("idle", sqrt(idleMaxSpeed));
+		}
+		if(requirePeaceful)
+			out.Write("peaceful");
+		if(optional)
+			out.Write("optional");
+		if(requireUncloaked)
+			out.Write("uncloaked");
+		if(repeatReset)
+			out.Write("repeat reset");
+		if(resetFired)
+			out.Write("reset fired");
+		if(resetCondition != ResetCondition::PAUSE)
+		{
+			static const map<ResetCondition, string> resets = {
+				{Timer::ResetCondition::NONE, "none"},
+				{Timer::ResetCondition::LEAVE_SYSTEM, "leave system"},
+				{Timer::ResetCondition::LEAVE_ZONE, "leave zone"}
+			};
+			auto it = resets.find(resetCondition);
+			if(it != resets.end())
+				out.Write("reset", it->second);
+		}
+		if(proximity > 0.)
+		{
+			string distance = "close";
+			if(!closeTo)
+				distance = "far";
+			if(proximityCenter)
+				out.Write("proximity", proximityCenter->Name());
+			else if(!proximityCenters.IsEmpty())
+			{
+				out.Write("proximity");
+				proximityCenters.Save(out);
+			}
+			out.Write("proximity settings", proximity, distance);
+		}
+
+		for(const auto &it : actions)
+			it.second.Save(out);
+
+	}
+	out.EndChild();
+}
+
+
+
+// Calculate the total time to wait, including any random value,
+// and instantiate the triggered action.
+Timer Timer::Instantiate(map<string, string> &subs,
+						const System *origin, int jumps, int64_t payload) const
+{
+	Timer result;
+	result.requireIdle = requireIdle;
+	result.requireUncloaked = requireUncloaked;
+	result.system = system;
+	result.systems = systems;
+	result.proximity = proximity;
+	result.proximityCenter = proximityCenter;
+	result.proximityCenters = proximityCenters;
+	result.closeTo = closeTo;
+	result.resetCondition = resetCondition;
+	result.repeatReset = repeatReset;
+	result.resetFired = resetFired;
+	result.idleMaxSpeed = idleMaxSpeed;
+	result.requirePeaceful = requirePeaceful;
+
+	result.timeElapsed = timeElapsed;
+	result.isComplete = isComplete;
+	result.isActive = isActive;
+
+	for(const auto &it : actions)
+		result.actions[it.first] = it.second.Instantiate(subs, origin, jumps, payload);
```

Comment:
MissionActions need to be validated during instantiation. (Copying code from Mission::Instantiate.)
```suggestion
	auto ait = actions.begin();
	for( ; ait != actions.end(); ++ait)
	{
		reason = ait->second.Validate();
		if(!reason.empty())
			break;
	}
	if(ait != actions.end())
	{
		Logger::LogError("Instantiation Error: Timer action \"" + TriggerToText(ait->first) + "\" in mission \""
			+ Identifier() + "\" uses invalid " + std::move(reason));
		return result;
	}
	for(const auto &it : actions)
		result.actions[it.first] = it.second.Instantiate(subs, sourceSystem, jumps, payload);
```